### ISSUE TITLE
Glm fails on snowSuperLearner but not mcSuperLearner

### ISSUE THREAD

**Issue by ck37** [ecpolley/SuperLearner#26]
Hello,
I'm hoping to use SuperLearner in parallel and ran into a weird case where glm models fail when called via snowSuperLearner but work just fine via mcSuperLearner. So I don't think it is a user error (apologies if it is). In both cases I am using 4 cores.
Here are my example calls to the two functions:
```r
testSNOW = snowSuperLearner(cluster = cl, Y = data$Y, X = x, SL.library = sl_lib,
                              method = "method.NNLS")
testMC = mcSuperLearner(Y = data$Y, X = x, SL.library = sl_lib, method = "method.NNLS")
```
In the snow version, SL generates a warning and the risk is NA for each glm model:
```r
Warning in snowSuperLearner(cluster = cl, Y = data$Y, X = x, SL.library =
sl_lib, : coefficients already 0 for all failed algorithm(s)
```
```r
testSNOW
## 
## Call:  
## snowSuperLearner(cluster = cl, Y = data$Y, X = x, SL.library = sl_lib,  
##     method = "method.NNLS") 
## 
## 
##                         Risk       Coef
## SL.glm.EstA_All           NA 0.00000000
## SL.glm.EstB_All           NA 0.00000000
## SL.glm.EstC_All           NA 0.00000000
## SL.glm.EstD_All           NA 0.00000000
## SL.ridge_All        3.997806 0.54550172
## SL.rpartPrune_All   4.852292 0.41289630
## SL.polymars_All    21.579445 0.04160199
## SL.mean_All       123.799548 0.00000000
```

I have attached a markdown test file & output, included R file, and the CSV data as a zip. I had to add .txt to the text files to satisfy github. This is on R 3.2.3 and SuperLearner 2.0-19 on OSX.

Thanks,
Chris

[RAssign3.csv.zip](https://github.com/ecpolley/SuperLearner/files/205543/RAssign3.csv.zip)
[sl-parallel.html.txt](https://github.com/ecpolley/SuperLearner/files/205546/sl-parallel.html.txt)
[sl-parallel.Rmd.txt](https://github.com/ecpolley/SuperLearner/files/205545/sl-parallel.Rmd.txt)
[RAssign3.Wrappers.R.txt](https://github.com/ecpolley/SuperLearner/files/205544/RAssign3.Wrappers.R.txt)

**Comment by ck37** [comment]
Sorry for the spam - after more testing I realized that I need to clusterExport() the custom GLM wrappers with SNOW, but not mcSuperLearner. User error.

**Comment by tedholzman** [comment]
HI.

I am trying to use snowSuperLearner on a fork cluster of 30 cpus.  As far as I can tell, using top, 30 cpus were allocated but only ten are being used.  Is there some default in snowSuperLearner that can be overridden?  Is there a hard limit of 10?  

Thanks.
     -Ted

**Comment by ck37** [comment]
Hi Ted,

Yes, unfortunately the current implementation parallelizes over the cross-validation folds, so it can only utilize as many cores as there are CV folds.

We do plan to improve this so that parallelization is over folds * algorithms, which will better leverage high core counts - I will try to get this done soonish. Another option would be to run SL sequentially but use multi-threaded versions of the learners that can take advantage of all 30 cores. Xgboost, ranger, and bartMachine are key algorithms where this can be done; glmnet could do this but it would require improving the wrapper. Example code: https://github.com/ck37/atlantic-causal-2017/blob/master/lib/ck_wrappers.R#L68

No biggie but in the future it would be best to open a new issue rather than append to a closed issue, unless you're having exactly the same problem as the original issue. In this case the title of this issue does not seem to reflect the problem you're reporting.

Thanks,
Chris
